compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -305,10 +327,14 @@ fn equate_impl_headers<'tcx>(
 /// of the two impls above to be empty.
 ///
 /// Importantly, this works even if there isn't a `impl !Error for MyLocalType`.
+///
+/// If there is no impossible obligation, this returns a list of obligations which
+/// overflowed by hitting the `recursion_limit` in the new solver. This is used
+/// to improve the error message.
 fn impl_intersection_has_impossible_obligation<'a, 'cx, 'tcx>(
     selcx: &mut SelectionContext<'cx, 'tcx>,
     obligations: &'a [PredicateObligation<'tcx>],
-) -> Option<PredicateObligation<'tcx>> {
+) -> Result<(), Vec<ty::Predicate<'tcx>>> {
```

Comment:
This logic is really confusing. Please turn this into an enum with proper states.

Reply:
the controlflow for `Err(vec![])` and `Err(vec![elements, inside])` is identical :thinking: changed it to a newtype instead